### ISSUE TITLE
Smooth trees issue

### ISSUE THREAD

**Issue by dudebro9800** [Cadiboo/NoCubes#75]
Despite the smooth option being toggled on, the trees still arent smooth; theyre blocky

**Comment by Cadiboo** [comment]
Sorry for the late response, can you post your client & server config and a screenshot please?

**Comment by dudebro9800** [comment]
Where can I locate my client and server config? I don't know what thiat is.
![2020-06-20_19 36 02](https://user-images.githubusercontent.com/64817077/85215441-6d7c8a00-b32d-11ea-8133-45dd8635c4c2.png)
![2020-06-20_19 35 49](https://user-images.githubusercontent.com/64817077/85215444-70777a80-b32d-11ea-9be2-eadd8bd98bf2.png)

**Comment by dudebro9800** [comment]
> Sorry for the late response, can you post your client & server config and a screenshot please?

hewllo?

**Comment by Cadiboo** [comment]
Hey, sorry, been a while. Your client config will be in "[.minecraft](https://minecraft.gamepedia.com/.minecraft#Locating_.minecraft)/config/nocubes-client.toml" and your server config will be in "[.minecraft](https://minecraft.gamepedia.com/.minecraft#Locating_.minecraft)/saves/world_name/serverconfig/nocubes-server.toml". Please also share your `latest.log` from "[.minecraft](https://minecraft.gamepedia.com/.minecraft#Locating_.minecraft)/logs/latest.log" and your `debug.log` (if it exists) from the same folder.

**Comment by dudebro9800** [comment]
![Screenshot (1)](https://user-images.githubusercontent.com/64817077/87217099-bd31ee00-c2fa-11ea-92e0-3d81502b61a8.png)
![Screenshot (2)](https://user-images.githubusercontent.com/64817077/87217100-bf944800-c2fa-11ea-9a69-9249b9a65e68.png)
![Screenshot (3)](https://user-images.githubusercontent.com/64817077/87217103-c15e0b80-c2fa-11ea-9e36-8472ea927113.png)


hope this helps dude or dudette

**Comment by dudebro9800** [comment]
> Hey, sorry, been a while. Your client config will be in "[.minecraft](https://minecraft.gamepedia.com/.minecraft#Locating_.minecraft)/config/nocubes-client.toml" and your server config will be in "[.minecraft](https://minecraft.gamepedia.com/.minecraft#Locating_.minecraft)/saves/world_name/serverconfig/nocubes-server.toml". Please also share your `latest.log` from "[.minecraft](https://minecraft.gamepedia.com/.minecraft#Locating_.minecraft)/logs/latest.log" and your `debug.log` (if it exists) from the same folder.

hello?

**Comment by Cadiboo** [comment]
Hey, really, really sorry that I've taken so long to help you with this.

You should just be able to change `renderSmoothAndVanillaLeaves` from "true" to "false" in your client config.

That option was added at someone else's request because it looked cool with their texture pack/shader.

It's taken so long to respond because:
- I've got a full time job and therefore don't have much time to work on my hobbies
- The computer I'm logged into GitHub on is at a different house from where I'm living + I can't go between them often because of covid
- I haven't worked on nocubes for a while because of the previous reasons and therefore didn't spot what should have been something obvious (plus didn't remember where the configs were stored)

I should have opened with this, but I'm a little bit more responsive on discord. Feel free to ping me on [the nocubes server](http://cadiboo.github.io/projects/nocubes/discord) in #support or pm me @Cadiboo#8887.

**Comment by dudebro9800** [comment]
> im [disappointed.] Despite toggling that off, the trees are still blocky. they still got that circular shape but with cubes showing like months ago when i complained
![2020-07-22_21 44 42](https://user-images.githubusercontent.com/64817077/88253016-17b93b80-cc65-11ea-8276-e1489fda17f2.png)
![2020-07-22_21 44 53](https://user-images.githubusercontent.com/64817077/88253017-18ea6880-cc65-11ea-8b85-673a0eb86297.png)
![2020-07-22_21 44 59](https://user-images.githubusercontent.com/64817077/88253018-1982ff00-cc65-11ea-9aee-d375d5e7a631.png)
![2020-07-22_21 45 02](https://user-images.githubusercontent.com/64817077/88253021-1ab42c00-cc65-11ea-9a0d-6587ab9473bb.png)

**Comment by Cadiboo** [comment]
Could you please send your client config and logs again? The fact that you aren’t getting the same issue with your terrain means that it’s got to be caused by that config setting being true or a bug in the mod. You change the value in client config to true when the game isn’t open and it should work when you next open the game. If it doesn’t work when you next open the game, there should be something like “correcting incorrect config value renderSmoothAndVanillaLeaves” that should show the issue.

**Comment by dudebro9800** [comment]
![Screenshot (6)](https://user-images.githubusercontent.com/64817077/88451349-d4003680-ce0a-11ea-9a14-0f2fdf06fa14.png)
Problem solved. But now i have a new problem: lack of leafiness despite the foliage mod is on. I toggled that config like you told me after quitting the game and restarting. But the leafiness doesn't appear like when my trees were blocky before that advice u gave me
![Screenshot (6)](https://user-images.githubusercontent.com/64817077/88451373-10339700-ce0b-11ea-824c-5d0a921e1523.png)

**Comment by Cadiboo** [comment]
Yeah, the extra foliage coming out from the sides of the leaves isn't in NoCubes, IIRC that's why the option to display both smooth and vanilla trees was added. NoCubes takes the texture from the side of the block and renders it bent, I'd have to write my own code to add that extra foliage (the way I've done with the short grass). This wouldn't be extremely hard, but I haven't done it yet.

**Comment by Cadiboo** [comment]
Hey, I’m working on updating the mod to 1.16 currently. I’ve decided to rewrite the mod from the ground up. I’m also planning on backporting the rewrite 1.12.2. Since I’m doing this I have the opportunity to implement support for fancy leaves right into nocubes without the pain of having to rewrite large amounts of code. If you could describe what you would like the leafiness of the trees to look like I‘ll work on implementing it :).

The purple things have been fixed in 1.16 already.

**Comment by dudebro9800** [comment]
> Hey, I’m working on updating the mod to 1.16 currently. I’ve decided to rewrite the mod from the ground up. I’m also planning on backporting the rewrite 1.12.2. Since I’m doing this I have the opportunity to implement support for fancy leaves right into nocubes without the pain of having to rewrite large amounts of code. If you could describe what you would like the leafiness of the trees to look like I‘ll work on implementing it :).
> 
> The purple things have been fixed in 1.16 already.

Yes thank you but I don’t want to play Minecraft in 1.16. I’m stuck on 1.12 because some of my fav mods haven’t updated beyond 1.12 and my favorite texture pack is at 1.12. 

In order to make the trees realistic, I use the foliage mod in conjunction with your mod . In spite of the foliage mod, the trees don’t look leafy while using your mod. If you notice in those pics from June 20 the trees were blocky and leafy — this was before I took your advice on how to un-blockify  the leaves. And they’re leafy because of the foliage mod. Now, after your advice, the leaves are no longer blocky but smooth; but smooth and not leafy like how I wanted them to be

**Comment by Cadiboo** [comment]
> Yes thank you but I don’t want to play Minecraft in 1.16.

I'm not asking you to.
> I’m stuck on 1.12 because some of my fav mods haven’t updated beyond 1.12 and my favorite texture pack is at 1.12.

I understand this, 1.12.2 was my favourite version as well, that's why I said "I’m also planning on backporting the rewrite 1.12.2."

> In order to make the trees realistic, I use the foliage mod in conjunction with your mod . In spite of the foliage mod, the trees don’t look leafy while using your mod.

Yeah, it's not possible for stuff like this to "just work" with the way NoCubes is written in 1.12.2. One of my goals in the rewrite is to make stuff more compatible.

> If you notice in those pics from June 20 the trees were blocky and leafy — this was before I took your advice on how to un-blockify the leaves. And they’re leafy because of the foliage mod. Now, after your advice, the leaves are no longer blocky but smooth; but smooth and not leafy like how I wanted them to be

You can change (back) your config, I've even created a config GUI for NoCubes.

**Comment by dudebro9800** [comment]
You done?

**Comment by Cadiboo** [comment]
With what?

**Comment by dudebro9800** [comment]
> With what?

what i wanted you to do

**Comment by Cadiboo** [comment]
I asked you what you wanted done
> I’ve decided to rewrite the mod from the ground up. I’m also planning on backporting the rewrite to 1.12.2. Since I’m doing this I have the opportunity to implement support for fancy leaves right into nocubes without the pain of having to rewrite large amounts of code. If you could describe what you would like the leafiness of the trees to look like I‘ll work on implementing it :).
> 
> The purple things have been fixed in 1.16 already.

And you replied
> Yes thank you but I don’t want to play Minecraft in 1.16. I’m stuck on 1.12 because some of my fav mods haven’t updated beyond 1.12 and my favorite texture pack is at 1.12.

So if you want me to implement leaves on trees as part of NoCubes then I’m still waiting for you to tell me what you’d like that to look like.

You also said
> If you notice in those pics from June 20 the trees were blocky and leafy — this was before I took your advice on how to un-blockify the leaves. And they’re leafy because of the foliage mod. Now, after your advice, the leaves are no longer blocky but smooth; but smooth and not leafy like how I wanted them to be

And I said that you could have
> - Vanilla style leaves (+ whatever the Foliage mod adds)
> - Smoothed NoCubes style leaves AND Vanilla style leaves (+ whatever the Foliage mod adds)
> - Just smoothed NoCubes style leaves

I also asked if by “the Foliage mod” you meant BetterFoliage. I asked this because BetterFoliage changes everything in a way that cannot be compatible with NoCubes (they intercept blocks mod-rendering and add extra details right into the raw buffer data that will be rendered to your screen).
If the mod is BetterFoliage I’ll need to implement the leaves from scratch myself the way I did with the short grass. If it’s not BetterFoliage I’ll happily look at how the extra leaves are added and try and add compatibility (I’ve done this before and the person thought that rendering them both at the same time would look good, so i added it).

**Comment by dudebro9800** [comment]
it's better foliage

also i dont understand what your point is. you dont know what i want? all i want are the trees to not be blocky. i want them to look like realistic trees so i used your mod and better foliage but it didsnt come out looking. and when i dont use the foliage mod but only yours the trees look too smooth and not like real trees and i tried to counteract this by using better foliage but that didnt work.

**Comment by Cadiboo** [comment]
This can’t be done in the mod currently. I’ll look into a solution. [This](https://simonschreibt.de/gat/airborn-trees/) looks promising and I should be able to do it without too much trouble by rendering the part of the model that isn’t the cube’s sides for each leaf block after I’ve smoothed the tree.
However, this won’t work with BetterFoliage because of how it does stuff. I think there’s a texture pack out there that does something similar though.
I’ll keep this issue open until it’s fixed.
It probably won’t be fixed soon though due to lack of free time.
